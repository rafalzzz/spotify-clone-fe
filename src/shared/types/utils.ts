export type TCustomRequest = {
  basicUrl?: string;
  endpoint?: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  requestBody?: Record<string, unknown>;
  allowCookies?: boolean;
};

export type MockResizeObserver = {
  observe: jest.Mock;
  unobserve: jest.Mock;
  disconnect: jest.Mock;
};
