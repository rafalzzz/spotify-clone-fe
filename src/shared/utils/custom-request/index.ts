import https from 'https';

import fetch from 'node-fetch';

const agent = new https.Agent({
  rejectUnauthorized: false,
});

type CustomRequestProps = {
  basicUrl?: string;
  endpoint?: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  requestBody?: Record<string, unknown>;
};

export const customRequest = ({
  basicUrl = process.env.NEXT_PUBLIC_API_URL,
  endpoint = '/',
  method,
  headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  requestBody,
}: CustomRequestProps) =>
  fetch(basicUrl + endpoint, {
    method,
    headers,
    agent,
    body: JSON.stringify(requestBody),
  });
