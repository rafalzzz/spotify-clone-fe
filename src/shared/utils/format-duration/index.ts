const pad = (number: number): string => {
  return ('0' + number).slice(-2);
};

export const formatDuration = (seconds: number, isReversedTime: boolean = false): string => {
  if (seconds < 0) return '0:00';

  const date = new Date(seconds * 1000);
  const hh = date.getUTCHours();
  const mm = date.getUTCMinutes();
  const ss = pad(date.getUTCSeconds());

  let value = `${mm}:${ss}`;

  if (hh) {
    value = `${hh}:${pad(mm)}:${ss}`;
  }

  if (isReversedTime) {
    return `-${value}`;
  }

  return value;
};
