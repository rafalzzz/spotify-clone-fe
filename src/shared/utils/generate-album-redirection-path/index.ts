export const generateAlbumRedirectionPath = (artistName: string): string =>
  `/album/${artistName.toLocaleLowerCase()}`;
