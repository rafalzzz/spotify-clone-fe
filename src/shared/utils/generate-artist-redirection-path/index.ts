export const generateArtistRedirectionPath = (artistName: string): string =>
  `/artist/${artistName.toLocaleLowerCase()}`;
