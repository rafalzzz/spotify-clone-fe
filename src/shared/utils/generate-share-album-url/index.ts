export const generateShareAlbumUrl = (albumId: number): string =>
  window.location.origin + `/album/${albumId}`;
