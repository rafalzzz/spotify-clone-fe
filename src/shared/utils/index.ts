export { convertMusicTrackToSongItem } from './convert-music-track-to-song-item';
export { getItemsAmount } from './get-items-amount';
export { getMainArtist } from './get-main-artist';
export { generateAlbumRedirectionPath } from './generate-album-redirection-path';
export { generateArtistRedirectionPath } from './generate-artist-redirection-path';
export { generateShareTrackUrl } from './generate-share-track-url';
export { generateTrackRedirectionPath } from './generate-track-redirection-path';
export { getShuffledIndexes } from './get-shuffled-indexes';
