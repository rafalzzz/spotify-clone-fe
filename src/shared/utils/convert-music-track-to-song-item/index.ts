import { TSongItem } from '@/types/components';
import { EMusicTrackKeys, TMusicTrack } from '@/types/music-track';

export const convertMusicTrackToSongItem = (musicTrack: TMusicTrack): TSongItem => ({
  [EMusicTrackKeys.ARTIST_NAME]: musicTrack[EMusicTrackKeys.ARTIST_NAME],
  [EMusicTrackKeys.COLLECTION_NAME]: musicTrack[EMusicTrackKeys.COLLECTION_NAME],
  [EMusicTrackKeys.TRACK_NAME]: musicTrack[EMusicTrackKeys.TRACK_NAME],
  [EMusicTrackKeys.PREVIEW_URL]: musicTrack[EMusicTrackKeys.PREVIEW_URL],
  [EMusicTrackKeys.ARTWORK_URL_60]: musicTrack[EMusicTrackKeys.ARTWORK_URL_60],
  [EMusicTrackKeys.ARTIST_ID]: musicTrack[EMusicTrackKeys.ARTIST_ID],
  [EMusicTrackKeys.COLLECTION_ID]: musicTrack[EMusicTrackKeys.COLLECTION_ID],
  [EMusicTrackKeys.TRACK_ID]: musicTrack[EMusicTrackKeys.TRACK_ID],
});
