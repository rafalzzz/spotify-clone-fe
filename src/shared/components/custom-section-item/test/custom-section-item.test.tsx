import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { generateAlbumRedirectionPath } from '@/utils/generate-album-redirection-path';


import { CustomSectionItem } from '../';

const REDIRECTION_TEST_ID = 'custom-section-item-redirection';

const MOCKED_CHILD_CONTENT = 'Mocked Child Content';
const MOCKED_COLLECTION_NAME = 'mocked collection name';
const MOCKED_CHILD = <div>{MOCKED_CHILD_CONTENT}</div>;

const renderCustomSectionItem = () =>
  render(
    <CustomSectionItem collectionName={MOCKED_COLLECTION_NAME}>{MOCKED_CHILD}</CustomSectionItem>,
  );

describe('CustomSectionItem', () => {
  it('renders component without error', () => {
    const screen = renderCustomSectionItem();
    expect(screen).toMatchSnapshot();
  });

  it('renders children correctly', () => {
    const { queryByText } = renderCustomSectionItem();
    expect(queryByText(MOCKED_CHILD_CONTENT)).toBeInTheDocument();
  });

  it('album redirection have correct path', () => {
    const { queryByTestId } = renderCustomSectionItem();

    const artistNameElement = queryByTestId(REDIRECTION_TEST_ID);
    expect(artistNameElement).toHaveAttribute(
      'href',
      generateAlbumRedirectionPath(MOCKED_COLLECTION_NAME),
    );
  });
});
