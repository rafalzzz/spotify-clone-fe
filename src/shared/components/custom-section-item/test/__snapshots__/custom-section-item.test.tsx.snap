// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSectionItem renders component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="custom-section-item__redirection"
        data-testid="custom-section-item-redirection"
        href="/album/mocked collection name"
      >
        <div
          class="custom-section-item"
        >
          <div
            class="custom-section-item__image"
          >
            <img
              alt="image"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=%2Fsome-image-url.jpg&w=256&q=75"
              srcset="/_next/image?url=%2Fsome-image-url.jpg&w=128&q=75 1x, /_next/image?url=%2Fsome-image-url.jpg&w=256&q=75 2x"
              style="color: transparent; height: 100%; width: 100%;"
              width="100"
            />
            <button
              class="custom-section-item__play-button"
              data-testid="custom-section-item-play-button"
            >
              <span
                aria-label="caret-right"
                class="anticon anticon-caret-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="caret-right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M715.8 493.5L335 165.1c-14.2-12.2-35-1.2-35 18.5v656.8c0 19.7 20.8 30.7 35 18.5l380.8-328.4c10.9-9.4 10.9-27.6 0-37z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div>
            Mocked Child Content
          </div>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="custom-section-item__redirection"
      data-testid="custom-section-item-redirection"
      href="/album/mocked collection name"
    >
      <div
        class="custom-section-item"
      >
        <div
          class="custom-section-item__image"
        >
          <img
            alt="image"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=%2Fsome-image-url.jpg&w=256&q=75"
            srcset="/_next/image?url=%2Fsome-image-url.jpg&w=128&q=75 1x, /_next/image?url=%2Fsome-image-url.jpg&w=256&q=75 2x"
            style="color: transparent; height: 100%; width: 100%;"
            width="100"
          />
          <button
            class="custom-section-item__play-button"
            data-testid="custom-section-item-play-button"
          >
            <span
              aria-label="caret-right"
              class="anticon anticon-caret-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="caret-right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M715.8 493.5L335 165.1c-14.2-12.2-35-1.2-35 18.5v656.8c0 19.7 20.8 30.7 35 18.5l380.8-328.4c10.9-9.4 10.9-27.6 0-37z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div>
          Mocked Child Content
        </div>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
